File: frontend/src/pages/Quiz.js
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Container, Paper, Typography, RadioGroup, Radio,
  FormControlLabel, Button, Box, LinearProgress,
  Alert, Grid, Chip, CircularProgress  
} from '@mui/material';
import { submitQuiz, getLoggedInUser } from '../api/requests';

const questionSets = {
  Beginner: [
    {
      id: 'vibe',
      question: "What kind of vibe are you going for?",
      type: 'single',
      options: ["Fresh and clean", "Warm and cosy", "Bold and attention-grabbing", "Light and subtle"]
    },
    {
      id: 'occasion',
      question: "Where are you most likely to wear this fragrance?",
      type: 'single',
      options: ["Daily wear (uni, errands)", "Special occasions (dinners, events)", "Outdoors (walks, picnics, travel)", "I just want to smell good at home"]
    },
    {
      id: 'season',
      question: "What season best fits your ideal scent?",
      type: 'single',
      options: ["Spring – floral, breezy", "Summer – fresh, citrusy", "Autumn – warm, spicy", "Winter – deep, woody"]
    },
    {
      id: 'scent_preference',
      question: "Do you prefer sweet or fresh scents?",
      type: 'single',
      options: ["Sweet (vanilla, caramel, fruity)", "Fresh (citrus, mint, green)", "I'm not sure yet"]
    },
    {
      id: 'longevity',
      question: "How long do you want the scent to last?",
      type: 'single',
      options: ["Just a few hours is fine", "I want it to last all day", "I don't mind either way"]
    }
  ],
  Intermediate: [
    {
      id: 'collection',
      question: "Which of these best describes your current fragrance collection?",
      type: 'single',
      options: ["A few designer scents", "Mostly fresh/clean perfumes", "Some warm and spicy options", "A little bit of everything"]
    },
    {
      id: 'scent_appeal',
      question: "Which type of scent do you find most appealing?",
      type: 'single',
      options: ["Fruity/floral", "Spicy/woody", "Aquatic/fresh", "Sweet/gourmand"]
    },
    {
      id: 'favorite_note',
      question: "Pick a note you know and love:",
      type: 'single',
      options: ["Vanilla", "Bergamot", "Leather", "Lavender"]
    },
    {
      id: 'strength',
      question: "How strong do you want your fragrance to be?",
      type: 'single',
      options: ["Subtle – just for me", "Noticeable – for people nearby", "Strong – I want to make an entrance"]
    },
    {
      id: 'brands',
      question: "Which of these brands have you tried or heard of?",
      type: 'single',
      options: ["Dior / Chanel / Versace", "Maison Margiela / Le Labo", "Mont Blanc / YSL / Paco Rabanne", "I'm still figuring them out"]
    }
  ],
  Advanced: [
    {
      id: 'top_notes',
      question: "Which of these top notes do you gravitate towards?",
      type: 'single',
      options: ["Bergamot", "Cardamom", "Pink Pepper", "Grapefruit"]
    },
    {
      id: 'base_notes',
      question: "Which base note do you enjoy most?",
      type: 'single',
      options: ["Amber", "Musk", "Oud", "Vetiver"]
    },
    {
      id: 'avoid',
      question: "Which of these would make you skip a fragrance?",
      type: 'single',
      options: ["Too much sweetness", "Poor projection", "Too powdery or soapy"]
    },
    {
      id: 'layering',
      question: "Do you enjoy layering fragrances?",
      type: 'single',
      options: ["Yes, I like to experiment", "Sometimes, depending on the scent", "No, I prefer one solid fragrance"]
    },
    {
      id: 'niche_brands',
      question: "What's your stance on niche brands?",
      type: 'single',
      options: ["Obsessed – I seek out niche houses", "Interested but still exploring", "Not interested"]
    }
  ]
};

const Quiz = () => {
  const navigate = useNavigate();
  const [currentStep, setCurrentStep] = useState(0);
  const [experienceLevel, setExperienceLevel] = useState(null);
  const [currentQuestion, setCurrentQuestion] = useState(0);
  const [answers, setAnswers] = useState({});
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  const [user, setUser] = useState(null);
  const [checkingAuth, setCheckingAuth] = useState(true);

  useEffect(() => {
    const checkAuth = async () => {
      try {
        const userData = await getLoggedInUser();
        setUser(userData);
      } catch (err) {
        console.error('Auth check error:', err);
      } finally {
        setCheckingAuth(false);
      }
    };
    checkAuth();
  }, []);

  const questions = questionSets[experienceLevel] || [];
  const progress = currentStep === 0 ? 20 : (20 + ((currentQuestion + 1) / questions.length) * 80);

  const handleExperienceSelect = (level) => {
    setExperienceLevel(level);
    setAnswers(prev => ({ ...prev, experience_level: level }));
    setCurrentStep(1);
  };

  const handleAnswer = (value) => {
    const question = questions[currentQuestion];
    setAnswers(prev => ({ ...prev, [question.id]: value }));
  };

  const handleNext = () => {
    if (currentQuestion < questions.length - 1) {
      setCurrentQuestion(prev => prev + 1);
      setError('');
    } else {
      handleSubmit();
    }
  };

  const handleBack = () => {
    if (currentQuestion > 0) {
      setCurrentQuestion(prev => prev - 1);
    } else {
      setCurrentStep(0);
    }
    setError('');
  };

  const handleSubmit = async () => {
    if (!user) {
      setError('Please log in to submit the quiz');
      return;
    }

    const unansweredQuestions = questions.filter(q => !answers[q.id]);
    if (unansweredQuestions.length > 0) {
      setError('Please answer all questions before getting recommendations');
      return;
    }

    setLoading(true);
    try {
      const quizData = {
        experience_level: answers.experience_level,
        longevity: answers.longevity,
        min_rating: 3.5,
        desired_accords: getDesiredAccords(answers),
        gender: "Unisex" // Ye NAYA FIELD ADD KARNA ZAROORI HAI
      };

      const response = await submitQuiz(quizData);

      if (!response.success) {
        throw new Error(response.error || 'Failed to get recommendations');
      }

      navigate('/recommendations', {
        state: {
          quizAnswers: answers,
          recommendations: response.recommendations
        }
      });
    } catch (err) {
      console.error("Quiz submission error:", err);
      setError(err.message || 'Failed to submit quiz. Please try again.');
    } finally {
      setLoading(false);
    }
  };

  // Helper function to map answers to scent accords
  const getDesiredAccords = (answers) => {
    const accords = [];
    const level = answers.experience_level;

    if (level === 'Beginner') {
      if (answers.vibe === "Fresh and clean") accords.push('Fresh', 'Clean', 'Aquatic');
      if (answers.season === "Summer") accords.push('Citrus', 'Fruity');
      // Add more mappings for beginner
    }
    else if (level === 'Intermediate') {
      if (answers.favorite_note === "Vanilla") accords.push('Vanilla', 'Gourmand');
      // Add more mappings for intermediate
    }
    else if (level === 'Advanced') {
      if (answers.top_notes === "Bergamot") accords.push('Citrus', 'Fresh');
      // Add more mappings for advanced
    }

    return [...new Set(accords)]; // Remove duplicates
  };

  const renderExperienceSelection = () => (
    <Box sx={{ mt: 4 }}>
      <Typography variant="h5" gutterBottom sx={{ color: '#FFD700' }}>
        How would you describe your experience with fragrances?
      </Typography>
      <Grid container spacing={2} sx={{ mt: 2 }}>
        {Object.keys(questionSets).map(level => (
          <Grid item xs={12} sm={4} key={level}>
            <Button
              fullWidth
              variant="outlined"
              size="large"
              onClick={() => handleExperienceSelect(level)}
              sx={{
                py: 3,
                borderColor: '#FFD700',
                color: '#FFD700',
                '&:hover': {
                  backgroundColor: 'rgba(255, 215, 0, 0.1)',
                },
              }}
            >
              <Typography variant="body1" sx={{ fontWeight: 'bold' }}>
                {level}
              </Typography>
              <Typography variant="caption" display="block">
                {level === 'Beginner' && 'Just starting out'}
                {level === 'Intermediate' && 'Tried a few fragrances'}
                {level === 'Advanced' && 'Understand notes and accords'}
              </Typography>
            </Button>
          </Grid>
        ))}
      </Grid>
    </Box>
  );

  const renderQuestion = () => {
    const question = questions[currentQuestion];
    const currentAnswer = answers[question.id];

    return (
      <Box sx={{ mt: 4 }}>
        <Typography variant="h5" gutterBottom sx={{ color: '#FFD700' }}>
          {question.question}
        </Typography>
        <RadioGroup
          value={currentAnswer || ''}
          onChange={(e) => handleAnswer(e.target.value)}
          sx={{ mt: 2 }}
        >
          {question.options.map((option) => (
            <FormControlLabel
              key={option}
              value={option}
              control={<Radio sx={{ color: '#FFD700' }} />}
              label={<Typography sx={{ color: 'white' }}>{option}</Typography>}
              sx={{
                mb: 1,
                '&:hover': { backgroundColor: 'rgba(255, 215, 0, 0.05)' },
                borderRadius: 1,
                padding: '4px 8px'
              }}
            />
          ))}
        </RadioGroup>
      </Box>
    );
  };

  if (checkingAuth) {
    return (
      <Container maxWidth="md" sx={{ py: 4 }}>
        <LinearProgress sx={{ color: '#FFD700' }} />
      </Container>
    );
  }

  if (!user) {
    return (
      <Container maxWidth="md" sx={{ py: 4 }}>
        <Alert severity="warning" sx={{ mb: 2, backgroundColor: '#2a2a2a', color: 'white' }}>
          Please log in to take the quiz and get personalized recommendations.
        </Alert>
        <Button
          variant="contained"
          onClick={() => navigate('/login')}
          sx={{
            backgroundColor: '#FFD700',
            color: 'black',
            '&:hover': { backgroundColor: '#e6c200' },
          }}
        >
          Go to Login
        </Button>
      </Container>
    );
  }

  return (
    <Container maxWidth="md" sx={{ py: 4 }}>
      <Paper elevation={3} sx={{
        p: 4,
        backgroundColor: '#1e1e1e',
        color: 'white',
        border: '1px solid #333'
      }}>
        <Typography variant="h4" gutterBottom align="center" sx={{ color: '#FFD700' }}>
          {currentStep === 0 ? 'Fragrance Experience Quiz' : 'Find Your Perfect Scent'}
        </Typography>

        <LinearProgress
          variant="determinate"
          value={progress}
          sx={{
            mb: 4,
            height: 8,
            borderRadius: 4,
            backgroundColor: '#333',
            '& .MuiLinearProgress-bar': { backgroundColor: '#FFD700' }
          }}
        />

        {currentStep === 0 ? renderExperienceSelection() : renderQuestion()}

        <Box sx={{ mt: 4, display: 'flex', justifyContent: 'space-between' }}>
          <Button
            variant="outlined"
            onClick={handleBack}
            disabled={(currentStep === 1 && currentQuestion === 0) || loading}
            sx={{
              borderColor: '#FFD700',
              color: '#FFD700',
              '&:hover': {
                backgroundColor: 'rgba(255, 215, 0, 0.1)',
              },
            }}
          >
            Back
          </Button>
          {currentStep === 1 && (
            <Button
              variant="contained"
              onClick={handleNext}
              disabled={!answers[questions[currentQuestion]?.id] || loading}
              sx={{
                backgroundColor: '#FFD700',
                color: 'black',
                '&:hover': { backgroundColor: '#e6c200' },
              }}
            >
              {loading ? (
                <CircularProgress size={24} sx={{ color: 'black' }} />
              ) : (
                currentQuestion === questions.length - 1 ? 'Get Recommendations' : 'Next'
              )}
            </Button>
          )}
        </Box>

        {error && (
          <Alert severity="error" sx={{
            mt: 2,
            backgroundColor: '#2a2a2a',
            color: 'white',
            border: '1px solid #ff3333'
          }}>
            {error}
          </Alert>
        )}
      </Paper>
    </Container>
  );
};

export default Quiz;

File: backend/routes/quiz.py
from flask import Blueprint, request, jsonify
import pandas as pd
import json
import ast  # New import for safer string evaluation
from models import db, QuizResult
from auth import login_required, get_current_user
import os

# Blueprint setup
quiz_bp = Blueprint('quiz', __name__)

# Dataset loading with error handling
def get_df():
    try:
        csv_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'perfume_data_clean.csv')
        df = pd.read_csv(csv_path)

        # Convert string representations to actual lists
        df['Main Accords'] = df['Main Accords'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else [])
        df['Perfumers'] = df['Perfumers'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else [])

        return df
    except Exception as e:
        print(f"Error loading dataset: {str(e)}")
        return pd.DataFrame()

# Experience-based question branching
def get_questions_by_level(experience_level):
    questions = {
        "Beginner": [
            {"id": "vibe", "question": "What kind of vibe are you going for?", "options": ["Fresh and clean", "Warm and cosy", "Bold and attention-grabbing", "Light and subtle"]},
            {"id": "occasion", "question": "Where will you wear this fragrance?", "options": ["Daily wear", "Special occasions", "Outdoors", "At home"]},
            {"id": "season", "question": "Preferred season for your scent?", "options": ["Spring", "Summer", "Autumn", "Winter"]},
            {"id": "scent_type", "question": "Sweet or fresh scents?", "options": ["Sweet", "Fresh", "Not sure"]},
            {"id": "longevity", "question": "How long should it last?", "options": ["Few hours", "All day", "Doesn't matter"]}
        ],
        "Intermediate": [
            {"id": "collection", "question": "Describe your current collection", "options": ["Few designer scents", "Mostly fresh/clean", "Some warm/spicy", "Bit of everything"]},
            {"id": "appeal", "question": "Most appealing scent type?", "options": ["Fruity/floral", "Spicy/woody", "Aquatic/fresh", "Sweet/gourmand"]},
            {"id": "note", "question": "Favorite note you know?", "options": ["Vanilla", "Bergamot", "Leather", "Lavender"]},
            {"id": "strength", "question": "How strong?", "options": ["Subtle", "Noticeable", "Strong"]},
            {"id": "brands", "question": "Brands you've tried?", "options": ["Dior/Chanel/Versace", "Maison Margiela/Le Labo", "Mont Blanc/YSL/Paco Rabanne", "Still figuring out"]}
        ],
        "Advanced": [
            {"id": "top_notes", "question": "Favorite top notes?", "options": ["Bergamot", "Cardamom", "Pink Pepper", "Grapefruit"]},
            {"id": "base_notes", "question": "Favorite base notes?", "options": ["Amber", "Musk", "Oud", "Vetiver"]},
            {"id": "avoid", "question": "What makes you skip a fragrance?", "options": ["Too sweet", "Poor projection", "Too powdery/soapy"]},
            {"id": "layering", "question": "Do you layer fragrances?", "options": ["Yes, I experiment", "Sometimes", "No, I prefer one"]},
            {"id": "niche", "question": "Your stance on niche brands?", "options": ["Obsessed", "Interested but exploring", "Not interested"]}
        ]
    }
    return questions.get(experience_level, [])

@quiz_bp.route('/quiz/start', methods=['POST'])
@login_required
def start_quiz():
    """Handle initial experience level selection"""
    user = get_current_user()
    data = request.json
    experience_level = data.get('experience_level')

    if not experience_level or experience_level not in ['Beginner', 'Intermediate', 'Advanced']:
        return jsonify({"error": "Valid experience level is required"}), 400

    # Save initial selection
    quiz_data = {"experience_level": experience_level}
    existing_result = QuizResult.query.filter_by(user_id=user.id).first()

    if existing_result:
        existing_result.preferences = json.dumps(quiz_data)
    else:
        new_result = QuizResult(user_id=user.id, preferences=json.dumps(quiz_data))
        db.session.add(new_result)

    try:
        db.session.commit()
        questions = get_questions_by_level(experience_level)
        return jsonify({"questions": questions}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({"error": str(e)}), 500

@quiz_bp.route('/quiz/submit', methods=['POST'])
@login_required
def submit_quiz():
    """Handle quiz submission and provide recommendations"""
    user = get_current_user()
    data = request.json
    answers = data.get('answers')

    if not answers:
        return jsonify({"error": "Answers are required"}), 400

    # Get existing quiz data
    result = QuizResult.query.filter_by(user_id=user.id).first()
    if not result:
        return jsonify({"error": "Quiz not started"}), 400

    try:
        preferences = json.loads(result.preferences)
        preferences.update(answers)

        # Save updated preferences
        result.preferences = json.dumps(preferences)
        db.session.commit()

        # Get recommendations
        df = get_df()
        if df.empty:
            return jsonify({"error": "Dataset not available"}), 500

        recommendations = get_recommendations(df, preferences)
        return jsonify({
            "recommendations": recommendations.to_dict(orient='records'),
            "message": "Recommendations based on your preferences"
        }), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({"error": str(e)}), 500

def get_recommendations(df, preferences):
    """Improved recommendation logic"""
    # Filter by experience level first
    exp_level = preferences.get('experience_level', 'Beginner')

    # Base filters
    filtered = df.copy()

    # Gender filter
    if preferences.get('gender'):
        filtered = filtered[filtered['Gender'] == preferences['gender']]

    # Rating filter
    min_rating = float(preferences.get('min_rating', 3.5))
    filtered = filtered[filtered['Rating Value'] >= min_rating]

    # Experience-specific filtering
    if exp_level == 'Beginner':
        # Simple filtering for beginners
        if preferences.get('vibe'):
            vibe_map = {
                'Fresh and clean': ['Fresh', 'Aquatic', 'Green'],
                'Warm and cosy': ['Vanilla', 'Amber', 'Gourmand'],
                'Bold and attention-grabbing': ['Spicy', 'Woody', 'Leather'],
                'Light and subtle': ['Floral', 'Citrus', 'Powdery']
            }
            desired_accords = vibe_map.get(preferences['vibe'], [])
            filtered['score'] = filtered['Main Accords'].apply(
                lambda accords: sum(1 for accord in accords if accord in desired_accords)
            )

    elif exp_level == 'Intermediate':
        # More advanced filtering
        if preferences.get('note'):
            filtered['score'] = filtered['Main Accords'].apply(
                lambda accords: 1 if preferences['note'] in accords else 0
            )

    else:  # Advanced
        # Most sophisticated filtering
        if preferences.get('top_notes') and preferences.get('base_notes'):
            filtered['top_score'] = filtered['Main Accords'].apply(
                lambda accords: 1 if preferences['top_notes'] in accords else 0
            )
            filtered['base_score'] = filtered['Main Accords'].apply(
                lambda accords: 1 if preferences['base_notes'] in accords else 0
            )
            filtered['score'] = filtered['top_score'] + filtered['base_score']

    # Sort and return top 5
    return filtered.sort_values('score', ascending=False).head(5)[[
        'Name', 'Gender', 'Rating Value', 'Rating Count',
        'Main Accords', 'Perfumers', 'Description', 'url'
    ]]
